### ISSUE TITLE
C# class implementing Java interface gives error CS5035 class Foo does not implement the interface member InterfaceName.<bridge>methodName()

### ISSUE THREAD

**Issue by martin-honnen** [ikvm-revived/ikvm#162]
Using IKVM 8.2.1 with .NET 6, I am trying to use C# to implement an abstract class in a Java library that also implements two interfaces in that library.
For some reason I don't understand at all currently, VS gives an error like "class Foo does not implement the interface member InterfaceName.<bridge>methodName()". Of course it offers me to generate the missing code but literally generates some syntactically incorrect construct
```
           ReturnType InterfaceName.<bridge>methodName()
            {
                throw new NotImplementedException();
            }
```

The Java interface doesn't have any methods using generics (or any mention of any `bridge` type).

I am kind of stuck where this lookup for such a method could come from and how I could resolve that or work around the problem.

I understand that probably nobody can help with that short description without seeing a full sample, unless the problem is somehow know.

So my question is, has anyone run into that problem about some spurious interface member with `<bridge>` being asked for to be implemented? Any ideas how to resolve it?

**Comment by martin-honnen** [comment]
To try to give some sample where the problem occurs, in a project having 
```
 <ItemGroup>
    <PackageReference Include="IKVM.Maven.Sdk" Version="1.0.1" />
    <MavenReference Include="net.sf.saxon:Saxon-HE" Version="11.4" />
  </ItemGroup>
```
add a class e.g.
```
using net.sf.saxon.om;
using net.sf.saxon.tree.iter;
using System.Xml;
using JArrayList = java.util.ArrayList;

namespace InterfaceProblemIKVMTest1
{
    sealed class AttributeEnumeration : AxisIterator, LookaheadIterator
    {

        private readonly JArrayList attList = new JArrayList(10); //ArrayList<XmlNode> attList = new ArrayList<>(10);
        private int ix = 0;
        private readonly DotNetNodeWrapper start;
        private DotNetNodeWrapper current;
        private DotNetDocumentWrapper docWrapper;
        private bool disposedValue;

        public AttributeEnumeration(DotNetDocumentWrapper docWrapper, DotNetNodeWrapper start)
        {
            this.docWrapper = docWrapper;
            this.start = start;
            XmlNamedNodeMap atts = start.node.Attributes;
            if (atts != null)
            {
                for (int i = 0; i < atts.Count; i++)
                {
                    string name = atts.Item(i).Name;
                    if (!(name.StartsWith("xmlns") &&
                            (name.Length == 5 || name[5] == ':')))
                    {
                        attList.add(atts.Item(i));
                    }
                }
            }
            ix = 0;
        }

        //@Override
        public bool supportsHasNext()
        {
            return true;
        }

        //@Override
        public bool hasNext()
        {
            return ix < attList.size();
        }

        //@Override
        public NodeInfo next()
        {
            if (ix >= attList.size())
            {
                return null;
            }
            current = makeWrapper((XmlNode)attList.get(ix), docWrapper, start, ix);
            ix++;
            return current;
        }

        Item SequenceIterator.next()
        {
            return ((AxisIterator)this).next();
        }

        public void close()
        {
            SequenceIterator.__DefaultMethods.close(this);
        }

        public void discharge()
        {
            SequenceIterator.__DefaultMethods.discharge(this);
        }

        private void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    // TODO: Verwalteten Zustand (verwaltete Objekte) bereinigen
                }

                // TODO: Nicht verwaltete Ressourcen (nicht verwaltete Objekte) freigeben und Finalizer überschreiben
                // TODO: Große Felder auf NULL setzen
                disposedValue = true;
            }
        }

        // // TODO: Finalizer nur überschreiben, wenn "Dispose(bool disposing)" Code für die Freigabe nicht verwalteter Ressourcen enthält
        // ~AttributeEnumeration()
        // {
        //     // Ändern Sie diesen Code nicht. Fügen Sie Bereinigungscode in der Methode "Dispose(bool disposing)" ein.
        //     Dispose(disposing: false);
        // }

        public void Dispose()
        {
            // Ändern Sie diesen Code nicht. Fügen Sie Bereinigungscode in der Methode "Dispose(bool disposing)" ein.
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        //Item AxisIterator.<bridge>next()
        //{
        //    throw new NotImplementedException();
        //}
    }
}
```

It will not compile (partially because it is not complete, the `DotNetNodeWrapper` is not provided), but it will show (at least for me here with VS 2022) the mentioned error in the form of "`AttributeEnumeration` doesn't implement the interface member `AxisIterator.<bridge>next()`".

**Comment by NightOwl888** [comment]
To answer your first question, this is the first time this issue has been reported AFAIK.

Out of curiosity, will it compile if you create a declaration as follows?

```cs

Item AxisIterator.next()
{
    throw new NotImplementedException();
}

```

**Comment by martin-honnen** [comment]
No, I get a `Compiler Error CS0539: 'AttributeEnumeration.next()' in explicit interface declaration is not a member of interface` in that case and some suggestion to "fix it" by inserting that `<bridge>` stuff again, although already in the suggestion popup that part has a red underline to indicate it doesn't compile.

**Comment by wasabii** [comment]
Well, the reason the name is mangled is because it's marked as ACC_BRIDGE in the .class file. Not sure I understand why this code would compile that way in Java, though. Bridge methods are general used for type erasure.

**Comment by wasabii** [comment]
It's most definitely not a pattern I'm familiar with from Java. It looks like the goal is to override and change the return type of an interface method.

from SequenceIterator:

```
/*@Nullable*/
Item next();
```

And now AxisIterator:

```
public interface AxisIterator extends SequenceIterator {
 
@Override
NodeInfo next();
 
}

```

@Override used on next(), with the same parameter signature, but different return type. The javac compiler appears to generate a default method on AxisIterator, marked with ACC_BRIDGE and ACC_SYNTHETIC, which implements SequenceIterator.next() and invokes this.next(), casting the result.

**Comment by martin-honnen** [comment]
@wasabii , thanks for looking into it. 

The various classes implementing one abstract class and additionally implementing one or two interfaces where sometimes the method names are the same and the return type is a subtype of the type used in another method has already given me a lot of headache, I couldn't find a way to rely on Java default implementation methods in the interfaces, it appears .NET doesn't recognize them but wants you to implement all interface methods; however, often I could fix that by doing some "reimplementation" like 
```
        public void close()
        {
            SequenceIterator.__DefaultMethods.close(this);
        }
```

On the issue of that <bridge> method, do you have any idea on how to get .NET to implement it so that the compilation is possible? I am still kind of lost, all three remaining compilation problems in my attempt to implement various classes show that odd <bridge> problem.

**Comment by wasabii** [comment]
So the magic here is that this is called a "covariant return type".

AxisIterator.next() overrides SequenceIterator.next(), but changes it's return type. This is allowed. And behind the scenes, a bridge method is generated that implements SequenceIterator.next() and does so by invoking AxisIterator.next().

This method is marked with ACC_BRIDGE and ACC_SYNTHETIC.

So how does IKVM handle this?

Well, in the CLR, at least until recently (lots of changes in Core need to be examined), you couldn't simply "delete" an interface method. You have to implement it. So, the old `Item next()` and `NodeItem next()` methods must be implemented by all classes extending from `AxisIterator`. IKVM solves this by actually implementing them automatically in concrete classes it is transpiling. So AxisIterator gets two interface methods. And then IKVM implements them BOTH in the classes that implement them.

But you're trying to implement the interface yourself, so IKVMC isn't around to do it for you. So you have to implement them both. But because they're meant to be hidden, the names are mangled, and you can't in C#.

Not sure how to address yet.

Well, I do have one solution: implement AxisIterator in java. That way ikvmc can generate the implementation for you. And then inherit from that in C#.

**Comment by wasabii** [comment]
@martin-honnen 

Would you be interested in trying out IKVM.NET.Sdk to do my recommendation? The idea is you can have actual .java source code in a .NET project.

So you could have a project named Saxon.Impl or some such, which contained actual .java files, one of which implemented AxisIterator as an abstract class, providing implementations for next(), which could be abstract, and inherited by your corresponding .NET code.

**Comment by martin-honnen** [comment]
@wasabii , I suppose that refers to https://www.nuget.org/packages/IKVM.NET.Sdk/8.2.2-prerelease0241#readme-body-tab? Yes, I will give it a try during this week. Can the Java code refer to .NET code, like ikvmstub compiled code could?

**Comment by wasabii** [comment]
Yup. It behaves like a real Net project. With references, nuget, Maven, etc. Prefix native .Net types with cli.

**Comment by wasabii** [comment]
Another possible solution: why don't we just not mangle the name? Mangling the bridge method name makes sense. But why the interface method?

**Comment by wasabii** [comment]
What do you mean exactly? As far as I can tell, covarient return types on interfaces isn't even available in C# 10.

**Comment by wasabii** [comment]
Consider the following code on .NET 6 C# 10:

```
    interface I1 { object M(); }
    interface I2 : I1 { int M(); }
```

This warns with CS0108 that I2.M() hides I1.M().

Additionally, attempting to implement I2 requires that both I1.M and I2.M be implemented by the consuming class.

This is different than in Java, where I2.M() will in fact override I1.M() with a new return type, and thus not require implementing classes to provide an implementation for I1.M().

**Comment by wasabii** [comment]
This is ultimately what IKVM generates:

```
    interface I1
    {

        object M();
    }

    interface I2 : I1
    {

        new int M();

        public static class __DefaultMethods
        {
            public static object M(I1 i) => i.M();

        }

    }

    class C : I2
    {

        public int M() => 0;

        object I1.M() => I2.__DefaultMethods.M(this);

    }
```

**Comment by NightOwl888** [comment]
This is what I mean:

```c#
    interface I1 { object M(); }
    interface I2 : I1 { int M(); }


    internal class C : I1, I2
    {
        object I1.M() => new object();
        int I2.M() => 0;
    }
```

Explicit interface declaration can be used to get different return types depending on the interface that the class instance is cast to.

```c#
            I1 x = new C();
            I2 y = x as I2;

            object o1 = x.M();
            int o2 = y.M();
```

Yes, the caller is expected to implement all interface members, unless using at least C# 8.0 where [default interface implementations are now allowed](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods).

However, for a situation like an enumerator, I don't see how that is an issue. All .NET collections implement multiple interfaces to override `GetEnumerator()` to return multiple enumerator types. So, it isn't unusual at all for this requirement to exist. The only thing that is unusual is the way IKVM seems to name the methods when there are multiple methods that have the same signature with a different return type. We can have multiple methods like this in the same class in C#, but all of them (or all of them but one) need to be explicitly declared to an interface when defining a class.

**Comment by wasabii** [comment]
So, just FYI, what you just described isn't "covariant return types". You had to implement the method with the same return type.

**Comment by NightOwl888** [comment]
Well, the docs say that default method implementations can be overridden, so that is a plus.

However, according to [this answer](https://stackoverflow.com/a/57020770), default interface members are not supported in .NET Framework. Sounds like more #ifdef statements to support such a thing on both Framework and Core.

**Comment by wasabii** [comment]
Right. So now you're caught up.

**Comment by martin-honnen** [comment]
@wasabii 
> 
> Would you be interested in trying out IKVM.NET.Sdk to do my recommendation? The idea is you can have actual .java source code in a .NET project.
> 

I have been trying to create a basic project but I fail to get anything buildable, VS 2022 for all dependencies just show problems through an exclamation mark and any attempt to build fails with an error `MSB4057	The target "ResolveIkvmExporter" does not exist in the project.	SaxonHE10Net6Test2IKVMSDK	C:\Users\marti\.nuget\packages\ikvm.net.sdk\8.2.2-prerelease0396\targets\IKVM.Java.Core.NoTasks.targets	11`.

The current setup is a project file

```
<Project Sdk="IKVM.NET.Sdk/8.2.2-prerelease0396">
  <PropertyGroup>
    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <Compile Remove="JavaClass1.java" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="IKVM.NET.Sdk" Version="8.2.2-prerelease0396" />
  </ItemGroup>

  <!--<ItemGroup>
	<PackageReference Include="IKVM.Maven.Sdk" Version="1.0.1" />
   </ItemGroup>

  <ItemGroup>
	<MavenReference Include="net.sf.saxon:Saxon-HE" Version="10.8" />
  </ItemGroup>-->

  <ItemGroup>
    <None Include="JavaClass1.java" />
  </ItemGroup>

</Project>	
```

Is there some sample project up somewhere on GitHub or elsewhere online?

**Comment by wasabii** [comment]
@martin-honnen First bug that fast, eh.

Add a property <UseIkvmTasks>true</UseIkvmTasks>

**Comment by wasabii** [comment]
Also you dont' need to include IKVM.NET.Sdk as a package reference.

**Comment by wasabii** [comment]
Okay. I misled you again. Do this in your project instead:
```
<Project>
    <PropertyGroup>
        <UseIkvmTasks>true</UseIkvmTasks>
    </PropertyGroup>
    <Import Sdk="IKVM.NET.Sdk" Version="8.2.2-prerelease0396" Project="Sdk.props"/>
    <PropertyGroup>
        <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
    </PropertyGroup>
    <ItemGroup>
        <None Include="JavaClass1.java" />
    </ItemGroup>
    <Import Sdk="IKVM.NET.Sdk" Version="8.2.2-prerelease0396" Project="Sdk.targets"/>
</Project>
```

**Comment by wasabii** [comment]
So I just did all of it locally and it worked. Made a IKVM.NET.Sdk project, added a AxisIteratorBase class, in .java:

```
import net.sf.saxon.om.NodeInfo;
import net.sf.saxon.tree.iter.AxisIterator;

public abstract class AxisIteratorBase implements AxisIterator
{

    @Override
    public abstract NodeInfo next();

}

```

And then made a C# project, added a project reference to the Java project, and created AxisIteratorImpl : AxisIteratorBase, and it only requires me to implement the single abstract method. All of the rest is set up properly.

**Comment by martin-honnen** [comment]
@wasabii , what is needed in these kind of IKVM.NET.Sdk projects to reference Saxon from Maven? Both
```
	<ItemGroup>
		<PackageReference Include="IKVM.Maven.Sdk" Version="1.0.1" />
		<MavenReference Include="net.sf.saxon:Saxon-HE" Version="11.4" />
	</ItemGroup>
```
or just
```
	<ItemGroup>
		<MavenReference Include="net.sf.saxon:Saxon-HE" Version="11.4" />
	</ItemGroup>
```

**Comment by wasabii** [comment]
The former. IKVM.Maven.Sdk isn't automatically part of IKVM.NET.Sdk.

**Comment by wasabii** [comment]
IKVM.NET.Sdk/8.2.2-prerelease0553 should have all those fixes applied, so no need to have UseIkvmTasks, or inline the SDK references.

**Comment by martin-honnen** [comment]
@wasabii , I am now trying to use 8.2.2-prerelease0553 with Java files of some larger project, any idea why I would get the following errors related to Java 8 JRE/JDK classes?
```
1>MSBUILD : error compiler.err.cant.access: cannot access java.util.Map.Entry
1>MSBUILD : error compiler.err.cant.access:   class file for java.util.Map$Entry not found
1>MSBUILD : error compiler.err.cant.access: cannot access java.util.stream.Collector.Characteristics
1>MSBUILD : error compiler.err.cant.access:   class file for java.util.stream.Collector$Characteristics not found
1>MSBUILD : error compiler.err.cant.resolve.location: cannot find symbol
1>MSBUILD : error compiler.err.cant.resolve.location:   symbol:   class Entry
1>MSBUILD : error compiler.err.cant.resolve.location:   location: interface java.util.Map
1>MSBUILD : error compiler.err.cant.access: cannot access javax.xml.datatype.DatatypeConstants.Field
1>MSBUILD : error compiler.err.cant.access:   class file for javax.xml.datatype.DatatypeConstants$Field not found
1>MSBUILD : error compiler.err.cant.resolve.location: cannot find symbol
1>MSBUILD : error compiler.err.cant.resolve.location:   symbol:   class Field
1>MSBUILD : error compiler.err.cant.resolve.location:   location: class javax.xml.datatype.DatatypeConstants
1>MSBUILD : error compiler.err.cant.access: cannot access java.util.Builder
1>MSBUILD : error compiler.err.cant.access:   bad class file: C:\.virtual-ikvm-home\assembly\IKVM.Java__8.2.0.0__13235d27fcbfff58\classes\java\util\Builder.class
1>MSBUILD : error compiler.err.cant.access:     unable to access file: File not found.
1>MSBUILD : error compiler.err.cant.access:     Please remove or make sure it appears in the correct subdirectory of the classpath.
```

msbuildproj has e.g.
```
<Project Sdk="IKVM.NET.Sdk/8.2.2-prerelease0553">
    <PropertyGroup>
        <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
    </PropertyGroup>
```

**Comment by wasabii** [comment]
Hmm. I do not. Just to verify I wasn't talking out my ass, I whipped up another quick IKVM.NET.Sdk project, and made sure it can still build stuff.... and it did.

The error you're seeing is the compiler trying to load the class stubs for various class files out of the VFS...... which shouldn't be environment specific in anyways. This is on Windows, right?

Any chance you can post your project file somewhere so I can look at it directly?

**Comment by martin-honnen** [comment]
I tried to reproduce it in a slightly different project, I don't get the exact same errors, but I do get one of those strange errors:
```
1>MSBUILD : error compiler.err.cant.access: cannot access java.util.Node
1>MSBUILD : error compiler.err.cant.access:   bad class file: C:\.virtual-ikvm-home\assembly\IKVM.Java__8.2.0.0__13235d27fcbfff58\classes\java\util\Node.class
1>MSBUILD : error compiler.err.cant.access:     unable to access file: File not found.
1>MSBUILD : error compiler.err.cant.access:     Please remove or make sure it appears in the correct subdirectory of the classpath.
1>Die Erstellung des Projekts "SaxonHE108J.msbuildproj" ist abgeschlossen -- FEHLER.

```

Online at https://github.com/martin-honnen/SaxonHE108JavaIkvmSdkTest3.

All done on Windows.

**Comment by wasabii** [comment]
Taking a look at it.

You really don't need to attempt to compile all of Saxon, though. All you needed was one class that implemented those two existing Saxon interfaces without the mangling.

**Comment by martin-honnen** [comment]
Yes, but I am trying to rebuild 10.8 for .NET Core anyway. In the end, if I have to write Java code, and as my original C# attempt tried to reimplement some stuff the Saxon folks had done in Java, it might be easier to do it myself in Java, using as much of the Java code from Saxon itself.

As for your help with the two existing interfaces, yes, your approach worked fine for that sample class/interface, unfortunately there is an outer class implementing an abstract class and two interfaces where I had the a similar `<bridge>` error and I couldn't resolve it the same way because that way my C# class then implemented two classes (and one interface) and that didn't work out.

**Comment by wasabii** [comment]
Alright, so, my guess is that's not an error, exactly.

Java has no notion of assemblies, remember. Just classpaths. Directories of .class files. Or jars which are just directories of class files.

The project has many errors with being unable to resolve packages: 
```
1>MSBUILD : error compiler.err.doesnt.exist: package net.sf.saxon.trans.packages does not exist
```

For instance. This error happens when it scans all the entries in the classpath and can't find find any such directory. But what if it DOES find a directory, but can't find the associated class in it? Then you get an error about the file not being found.

Basically, somewhere in Saxon, it is referencing a type "Node". And in that file, you have a series of wildcard "import" statements. And so the compiler is searching in each possible one for any file named "Node.class". Just checking each one, in order, for the class.

So you have a file somewhere that has two import statements: one where Node SHOULD BE, but isn't, and the other being java.util.*. Probably near the end. So it's not finding Node where it should find it, and so it's continuing on and checking every other wildcard import, and java.util.* is the last one.

It's probably DocumentWrapper.java:

```
    public final Node docNode;
```

And at the top:
```

package net.sf.saxon.dom;

import net.sf.saxon.Configuration;
import net.sf.saxon.functions.ResolveURI;
import net.sf.saxon.lib.NamespaceConstant;
import net.sf.saxon.om.AxisInfo;
import net.sf.saxon.om.GenericTreeInfo;
import net.sf.saxon.om.NodeInfo;
import net.sf.saxon.pattern.NodeKindTest;
import net.sf.saxon.tree.iter.AxisIterator;
import org.w3c.dom.*;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.*;
```

My guess is org.w3c.dom is missing. And Node should be in there. And java.util.* is at the end.

**Comment by wasabii** [comment]
I am still convinced you could solve this easier using the bridge class method. If you can write the code in C#, you can write it in Java.

**Comment by martin-honnen** [comment]
But `org.w3c.dom.Node` is part of Java 8 https://docs.oracle.com/javase/8/docs/api/org/w3c/dom/Node.html, shouldn't it be available to IKVM based on OpenJDK Java 8/1.8?

**Comment by wasabii** [comment]
Hmm. You're right.

I did however add `import org.w3c.dom.Node` to the top of the file and the error went away. So yeah, this is probably some bug of some additional bug of some sort somewhere with fake classpaths.

**Comment by martin-honnen** [comment]
As for things being easier with the bridge class method, as I said, I couldn't get it work with some other class, namely trying to reimplement https://saxonica.plan.io/projects/saxonmirrorhe/repository/he/revisions/he-saxon10_6/entry/src/main/java/net/sf/saxon/dotnet/DotNetNodeWrapper.java (which is for Saxon 10) for Saxon 11 but in C# as there I run into another bridge mangling problem but your resolution for the previous one, namely implementing one of the interfaces as an abstract class in Java to be inherited in C#, fails for me as the class DotNetNodeWrapper.cs would then inherit from two classes.

**Comment by wasabii** [comment]
Huh. Is that from their official C# library? I didn't know they used IKVM.

Could you just copy their DotNetNodeWrapper code into IKVM.Net.Sdk, and use that? But keep using the rest of Saxon from Maven?

**Comment by martin-honnen** [comment]
The original Saxon HE (well PE and EE as well) .NET product indeed uses the original IKVM 8.1 for .NET framework.
But they stopped using that as it didn't support .NET core and they pioneered some inhouse Java to C# transpilation (based on XSLT 3 Java syntax tree to C# syntax tree transformations) for their new SaxonCS , .NET 5/.NET core product line.

As for using that DotNetNodeWrapper, yes, it works fine for me with Saxon HE 10, although my initial attempts were directed towards Saxon HE 11, as that is currently supported main release. 

I have Saxon HE 10.8 running under .NET 3.1 core thanks to your IKVM/IKVM.SDK (https://github.com/martin-honnen/SaxonHE10Net31), but it identifies itself as the Java product, as there is some Platform switch they turn on for the .NET product, it seems, that requires recompiling everything through IKVM, at least as far as I have figured it out so far. And on that way I have today run into those odd error messages I asked about earlier. I will need to work through them step by step, I guess, to see, whether some more explicit import, like your `import org.w3c.dom.Node` makes them vanish.

 As for "Could you just copy their DotNetNodeWrapper code into IKVM.Net.Sdk, and use that? But keep using the rest of Saxon from Maven?", I have partly achieved that, only the platform identification does appear to require me to recompile everything through IKVM.

**Comment by wasabii** [comment]
Ahh. It's Version.java. Has a bunch of hard coded stuff to identify the platform.

So, my goal with all this Maven/IkvmReference stuff, is so that Java developers don't have to do their own work by adding .NET stuff into their Java code. They should just publish Java code. And if they have some .NET wrappers or extension methods or other niceties, they can make a separately maintained NuGet package that builds ontop of their Maven package.

So, ideally, all the .NET stuff should come out of Saxon. It shouldn't care about the platform.

They have some lookup thing on their Platform interface, which has a isJava or isDotNet method. But I don't see it actually called from anywhere, so maybe it doesn't matter?

**Comment by wasabii** [comment]
Ultimately it shouldn't be Saxon deciding what JVM they're running on. That stuff comes from System properties, which IKVM does set properly already.
